text/0091-dht-record-creation-time.md

```diff
@@ -0,0 +1,105 @@
+
+The authority discovery DHT records use the protobuf protocol and the current format is specified [here](https://github.com/paritytech/polkadot-sdk/blob/313fe0f9a277f27a4228634f0fb15a1c3fa21271/substrate/client/authority-discovery/src/worker/schema/dht-v2.proto#L4). This RFC proposese extending the schema in a backwards compatible manner by adding a new optional `creation_time` field to `AuthorityRecord` and nodes can use this information to determine which of the record is newer.
+
+Diff of `dht-v3.proto` vs `dht-v2.proto`
+
+```
+@@ -1,10 +1,10 @@
+ syntax = "proto3";
+
+-package authority_discovery_v2;
++package authority_discovery_v3;
+
+ // First we need to serialize the addresses in order to be able to sign them.
+ message AuthorityRecord {
+ 	repeated bytes addresses = 1;
++	// Time since UNIX_EPOCH in nanoseconds, scale encoded
++	TimestampInfo creation_time = 2;
+ }
+
+ message PeerSignature {
+@@ -13,11 +15,17 @@
+ 	bytes public_key = 2;
+ }
+
++// Information regarding the creation data of the record
++message TimestampInfo {
++       // Time since UNIX_EPOCH in nanoseconds, scale encoded
++       bytes timestamp = 1;
++}
++
+```
+
+Each time a node wants to resolve an authorithy ID it will issue a query with a certain redundancy factor, and from all the results it receives it will decide to pick only the newest record. Additionally, the nodes that answer with old records will be updated with the newer record.
+
+
+## Drawbacks
+
+In theory the new protocol creates a bit more traffic on the DHT network, because it waits for DHT records to be received from more than one node, while in the current implementation we just take the first record that we receive and cancel all in-flight requests to other peers. However, because the redundancy factor will be relatively small and this operation happens rarerly, every 10min, this cost is negligible.
+
+## Testing, Security, and Privacy
+
+
+This RFC's implementation https://github.com/paritytech/polkadot-sdk/pull/3786 had been tested on various local test networks and versi.
+
+With regard to security the creation time is wrapped inside SignedAuthorityRecord wo it will be signed with the authority id key, so there is no way for other malicious nodes to manipulate this field without the received node observing.
+
+## Performance, Ergonomics, and Compatibility
+
+Irrelevant.
+
+### Performance
+
+Irrelevant.
+
+### Ergonomics
+
+Irrelevant.
+
+### Compatibility
+
+The changes are backwards compatible with the existing protocol, so nodes with both the old protocol and newer protocol can exist in the network, this is achieved by the fact that we use protobuf for serializing and deserializing the records, so new fields will be ignore when deserializing with the older protocol and vice-versa when deserializing an old record with the new protocol the new field will be `None` and the new code accepts this record as being valid.
```

Comment:
How does that works on signature verification? Do we use the raw encoded protobuf bytes? Because the signature also needs to sign the creation time?

Reply:
Yes, we first serialize the AuthorityRecord then  we signed it, see here: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/client/authority-discovery/src/worker.rs#L447, so by adding new fields to AuthorithyRecord those bytes will be part of the signed payload.

And vice-versa if you get a record without this field, the signature we will be valid and when we deserialize it the creation time will be None, so you can concur this are nodes running the old version.

Reply:
But a node that doesn't have upgraded, will not know about the creation time. So, it will not decode it and thus, fail to check the signature.

So, this is RFC is actually breaking and would require all collators etc to upgrade.

So, they idea of having a second signature is actually better. 🙈

Reply:
I was looking at the wrong code! Everything is correct as far as I can say 😬